casper/src/test/scala/coop/rchain/casper/util/comm/FairRoundRobinDispatcherSpec.scala

```diff
@@ -0,0 +1,589 @@
+              _ <- dispatcher.dispatch("B", 1)
+              _ <- dispatcher.dispatch("B", 2)
+              _ <- dispatcher.dispatch("C", 3)
+            } yield ()
+          )(
+            expectedQueue = Queue("A", "B", "C"),
+            expectedMessages = Map(
+              "A" -> Queue.empty[Int],
+              "B" -> Queue(1, 2),
+              "C" -> Queue(3)
+            ),
+            expectedRetries = Map(
+              "A" -> 0,
+              "B" -> 0,
+              "C" -> 0
+            ),
+            expectedSkipped = 3
+          )
+        }
+
+      "enqueue when waiting for message from other sources and drop if full" in
+        new TestEnv(2, 10, 10) {
+          validate(
+            for {
+              _ <- dispatcher.ensureSourceExists("A")
+              _ <- dispatcher.ensureSourceExists("B")
+              _ <- dispatcher.ensureSourceExists("C")
+              _ <- dispatcher.dispatch("B", 1)
+              _ <- dispatcher.dispatch("B", 2)
+              _ <- dispatcher.dispatch("C", 3)
+              _ <- dispatcher.dispatch("B", 4)
+            } yield ()
+          )(
+            expectedQueue = Queue("A", "B", "C"),
+            expectedMessages = Map(
+              "A" -> Queue.empty[Int],
+              "B" -> Queue(1, 2),
+              "C" -> Queue(3)
+            ),
+            expectedRetries = Map(
+              "A" -> 0,
+              "B" -> 0,
+              "C" -> 0
+            ),
+            expectedSkipped = 4
+          )
+        }
+
+      "dispatch as many messages from buffer as possible" in
+        new TestEnv(10, 10, 10) {
+          validate(
+            for {
+              _ <- dispatcher.ensureSourceExists("A")
+              _ <- dispatcher.ensureSourceExists("B")
+              _ <- dispatcher.ensureSourceExists("C")
+              _ <- dispatcher.dispatch("B", 1)
+              _ <- dispatcher.dispatch("B", 2)
+              _ <- dispatcher.dispatch("C", 3)
+              _ <- dispatcher.dispatch("A", 4)
+            } yield ()
+          )(
+            expectedQueue = Queue("A", "B", "C"),
+            expectedMessages = Map(
+              "A" -> Queue.empty[Int],
+              "B" -> Queue(2),
+              "C" -> Queue.empty[Int]
+            ),
+            expectedRetries = Map(
+              "A" -> 0,
+              "B" -> 0,
+              "C" -> 0
+            ),
+            expectedHandled = Queue("A" -> 4, "B" -> 1, "C" -> 3)
+          )
+        }
+
+      "give up after skipped messages and continue with other sources" in
+        new TestEnv(10, 3, 10) {
+          validate(
+            for {
+              _ <- dispatcher.ensureSourceExists("A")
+              _ <- dispatcher.ensureSourceExists("B")
+              _ <- dispatcher.ensureSourceExists("C")
+              _ <- dispatcher.dispatch("B", 1)
+              _ <- dispatcher.dispatch("B", 2)
+              _ <- dispatcher.dispatch("C", 3)
+            } yield ()
```

Comment:
Very nice work with this PR. :)

One side question. Is there a reason why not use just an applicative in these places? It seems less noisy to me.

```scala
validate(
  dispatcher.ensureSourceExists("A") *>
    dispatcher.ensureSourceExists("B") *>
    dispatcher.ensureSourceExists("C") *>
    dispatcher.dispatch("B", 1) *>
    dispatcher.dispatch("B", 2) *>
    dispatcher.dispatch("C", 3)
)
```

Reply:
Thanks, this is exactly the answer I was looking for.